src/spaced/spaced/game/weapons/gun_beam.hpp

```diff
@@ -12,7 +12,7 @@ class GunBeam final : public Weapon {
 		float dps{1.0f};
 	};
 
-	explicit GunBeam(Services const& services) : Weapon(services, "DeathRay") {}
+	explicit GunBeam(Services const& services);
```

Comment:
why pass a reference to `services` here, and in all the other places? could you also have some `ServiceLocator` with services and include it in any needed `.cpp`s? if not, why not?

Reply:
`Services` *is* a service locator, it's just not globally accessible / a singleton, it has its own lifetime and ownership, which supercedes all the services it owns, thereby giving them also explicit lifetime and ownership. Globals, especially singletons, take away both of these very important aspects in C++.

The rest of your thinking is spot on: instead of passing dozens of different references in various constructors (which would be most explicit and correct, but too cumbersome and volatile), we design commonly referenced items as services, initialized by the game driver (`Spaced`).